Code/network/src/Resolver.cpp

```diff
@@ -0,0 +1,148 @@
+#include "Resolver.h"
+
+#include <utility>
+
+Resolver::Resolver(const std::string & acAddress):
+    m_futureEndpoints(std::future<std::vector<Endpoint>>()),
+    m_endpoints(std::vector<Endpoint>())
+{
+    Parse(acAddress);
+}
+
+Resolver::Resolver(Resolver && aRhs) noexcept:
+    m_futureEndpoints{std::move(aRhs.m_futureEndpoints)},
+    m_endpoints{std::move(aRhs.m_endpoints)}
+{
+}
+
+size_t Resolver::GetSize() noexcept
+{
+    return GetEndpoints().size();
+}
+
+bool Resolver::IsEmpty() noexcept
+{
+    return GetSize() == 0;
+}
+
+Endpoint Resolver::GetEndpoint(const size_t & acIndex) noexcept
+{
+    return GetEndpoints().at(acIndex);
+}
+
+Endpoint Resolver::operator[](const size_t & acIndex) noexcept
+{
+    return GetEndpoint(acIndex);
+}
+
+Resolver & Resolver::operator=(Resolver && aRhs) noexcept
+{
+    m_endpoints = std::move(aRhs.m_endpoints);
+    m_futureEndpoints = std::move(aRhs.m_futureEndpoints);
+
+    return *this;
+}
+
+Resolver::Iterator Resolver::begin() noexcept
+{
+    return Iterator(GetEndpoints().cbegin());
+}
+
+Resolver::Iterator Resolver::end() noexcept
+{
+    return Iterator(GetEndpoints().cend());
+}
+
+std::vector<Endpoint> & Resolver::GetEndpoints() noexcept
+{
+    if (m_futureEndpoints.valid()) {
+        // std::future::get can only be called once in its lifetime, so we have to keep them all in m_endpoints
+        const std::vector<Endpoint> & newEndpoints = m_futureEndpoints.get();
+        m_endpoints.insert(std::end(m_endpoints), std::begin(newEndpoints), std::end(newEndpoints));
+    }
+
+    return m_endpoints;
+}
+
+void Resolver::Parse(const std::string & acAddress) noexcept
+{
+    uint16_t port = 0;
+    std::string strAddress(acAddress);
```

Comment:
Might as well just pass the argument by value, this way we can get a move optimization

Reply:
Exactly my thought. I just copied it from your code and preferred not to touch it. Glad that you want it fixed 😄